src/lib/logical_query_plan/lqp_unique_constraint.hpp

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "expression/abstract_expression.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * Container structure to define uniqueness for subsets of LQP output expressions. Analogous to SQL's UNIQUE
+ * constraint, rows containing NULL values in any of the expressions are always considered to be distinct. For
+ * PRIMARY KEY semantics, check if the expressions are nullable, cf. AbstractLQPNode::is_column_nullable.
+ *
+ * NOTE: Unique constraints are only valid for LQP nodes that contain no invalidated rows (i.e., where there has
+ *       been a ValidateNode before or where MVCC is disabled).
```

Comment:
@julianmenzler This is a very important warning as people might do illegal PQP optimizations if they solely rely on the unique constraint without having validated the table. Any idea how to enforce this?

Reply:
Good catch!
I do not see an elegant solution. But the following idea might be at least something we can do:

* For StoredTableNode and StaticTableNode return an empty vector for `unique_constraints()`.
* Add a private function (e.g. `_unique_constraints()`) that is visible to ValidateNode only, returning the actual constraints.

However, the above solution breaks unique constraints for LQPs when ValidateNodes become moved up in the tree. Is this ever the case?

Reply:
I don't think that we should hide this. The super-duper-index-scan-with-validation-looking-from-the-back-stopping-at-the-first-result-thing needs that information and will perform the validation internally. It is not *wrong* to use the information on non-validated nodes, it is just really dangerous to perform optimizations in the operator if you are not aware of the MVCC limitation.

Reply:
I agree. Then we should leave it as is, I guess.